src/py21cmfast/src/Stochasticity.c

```diff
@@ -213,22 +213,90 @@ void stoc_set_consts_cond(struct HaloSamplingConstants *const_struct, double con
     return;
 }
 
-void place_on_hires_grid(int x, int y, int z, int *crd_hi, gsl_rng *rng) {
-    // we want to randomly place each halo within each lores cell,then map onto hires
-    // this is so halos are on DIM grids to match HaloField and Perturb options
-    int x_hi, y_hi, z_hi;
+// TODO: perhaps some of these functions should be moved to an indexing.c file
+// Minutia: I'm guessing that since particles will rarely be wrapped, the branching
+//  will be faster than a modulus, either way this is not likely to be a bottleneck
+void wrap_position(double pos[3], double size[3]) {
+    // wrap the coordinates to the box size
+    while (pos[0] >= size[0]) {
+        pos[0] -= size[0];
+    }
+    while (pos[0] < 0) {
+        pos[0] += size[0];
+    }
+    while (pos[1] >= size[1]) {
+        pos[1] -= size[1];
+    }
+    while (pos[1] < 0) {
+        pos[1] += size[1];
+    }
+    while (pos[2] >= size[2]) {
+        pos[2] -= size[2];
+    }
+    while (pos[2] < 0) {
+        pos[2] += size[2];
+    }
+    return;
+}
+
+void wrap_coord(int idx[3], int size[3]) {
+    // wrap the coordinates to the box size
+    while (idx[0] >= size[0]) {
+        idx[0] -= size[0];
+    }
+    while (idx[0] < 0) {
+        idx[0] += size[0];
+    }
+    while (idx[1] >= size[1]) {
+        idx[1] -= size[1];
+    }
+    while (idx[1] < 0) {
+        idx[1] += size[1];
+    }
+    while (idx[2] >= size[2]) {
+        idx[2] -= size[2];
+    }
+    while (idx[2] < 0) {
+        idx[2] += size[2];
+    }
+    return;
+}
+
+void random_point_in_sphere(double centre[3], double radius, gsl_rng *rng, double *point) {
+    // generate a random point in a sphere of given radius and centre
+    double x1, y1, z1, d1, r1;
+    x1 = gsl_rng_uniform(rng);
+    y1 = gsl_rng_uniform(rng);
+    z1 = gsl_rng_uniform(rng);
+    d1 = sqrt(x1 * x1 + y1 * y1 + z1 * z1);
+    r1 = gsl_rng_uniform(rng);
+    point[0] = centre[0] + (radius * r1 * x1 / d1);
+    point[1] = centre[1] + (radius * r1 * y1 / d1);
+    point[2] = centre[2] + (radius * r1 * z1 / d1);
```

Comment:
How can this give negative values of a coordinate? `radius` is positive, `r1` is on (0,1), `x1` is on (0,1) and d1 is positive.